File: app/ios/Runner/DualCameraPlugin.swift
import Flutter
import UIKit
import AVFoundation

/// Dual-camera recording plugin for iOS
/// Uses AVCaptureMultiCamSession for simultaneous front + back camera recording
/// Requires iPhone XS or later, iOS 13+
class DualCameraPlugin: NSObject, FlutterPlugin {
    private var registrar: FlutterPluginRegistrar
    private var eventSink: FlutterEventSink?

    // Multi-cam session
    private var multiCamSession: AVCaptureMultiCamSession?

    // Camera inputs
    private var frontCameraInput: AVCaptureDeviceInput?
    private var backCameraInput: AVCaptureDeviceInput?

    // Video outputs
    private var frontVideoOutput: AVCaptureMovieFileOutput?
    private var backVideoOutput: AVCaptureMovieFileOutput?

    // Audio
    private var audioOutput: AVCaptureAudioDataOutput?
    private var audioInput: AVCaptureDeviceInput?

    // Texture renderers for preview
    private var frontTextureEntry: FlutterTextureEntry?
    private var backTextureEntry: FlutterTextureEntry?

    // Recording state
    private var isRecording = false
    private var sessionDir: String?
    private var sessionId: String?
    private var recordingStartTime: Date?
    private var isSwapped = false

    // Audio level monitoring
    private var audioLevelTimer: Timer?

    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        super.init()
    }

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.stori/dual_camera",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "com.stori/dual_camera_events",
            binaryMessenger: registrar.messenger()
        )

        let instance = DualCameraPlugin(registrar: registrar)
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isDualCameraSupported":
            result(AVCaptureMultiCamSession.isMultiCamSupported)

        case "initialize":
            initialize(call: call, result: result)

        case "startRecording":
            startRecording(call: call, result: result)

        case "pauseRecording":
            pauseRecording(result: result)

        case "resumeRecording":
            resumeRecording(result: result)

        case "stopRecording":
            stopRecording(result: result)

        case "swapCameras":
            isSwapped = !isSwapped
            result(nil)

        case "dispose":
            cleanup()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Initialize

    private func initialize(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard AVCaptureMultiCamSession.isMultiCamSupported else {
            result(FlutterError(code: "UNSUPPORTED", message: "Multi-cam not supported on this device", details: nil))
            return
        }

        do {
            // Configure audio session for voice recording
            let audioSession = AVAudioSession.sharedInstance()
            try audioSession.setCategory(.playAndRecord, mode: .voiceChat, options: [.defaultToSpeaker, .allowBluetooth])
            try audioSession.setActive(true)

            // Create multi-cam session
            let session = AVCaptureMultiCamSession()
            session.beginConfiguration()

            // Front camera
            guard let frontCamera = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .front) else {
                result(FlutterError(code: "NO_CAMERA", message: "Front camera not available", details: nil))
                return
            }
            let frontInput = try AVCaptureDeviceInput(device: frontCamera)
            if session.canAddInput(frontInput) {
                session.addInputWithNoConnections(frontInput)
            }

            // Back camera
            guard let backCamera = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back) else {
                result(FlutterError(code: "NO_CAMERA", message: "Back camera not available", details: nil))
                return
            }
            let backInput = try AVCaptureDeviceInput(device: backCamera)
            if session.canAddInput(backInput) {
                session.addInputWithNoConnections(backInput)
            }

            // Audio input
            guard let audioDevice = AVCaptureDevice.default(for: .audio) else {
                result(FlutterError(code: "NO_MIC", message: "Microphone not available", details: nil))
                return
            }
            let audioDeviceInput = try AVCaptureDeviceInput(device: audioDevice)
            if session.canAddInput(audioDeviceInput) {
                session.addInputWithNoConnections(audioDeviceInput)
            }

            // Video outputs
            let frontOutput = AVCaptureMovieFileOutput()
            let backOutput = AVCaptureMovieFileOutput()

            if session.canAddOutput(frontOutput) {
                session.addOutputWithNoConnections(frontOutput)
            }
            if session.canAddOutput(backOutput) {
                session.addOutputWithNoConnections(backOutput)
            }

            // Connect front camera to front output
            if let frontVideoPort = frontInput.ports(for: .video, sourceDeviceType: frontCamera.deviceType, sourceDevicePosition: .front).first {
                let frontConnection = AVCaptureConnection(inputPorts: [frontVideoPort], output: frontOutput)
                if session.canAddConnection(frontConnection) {
                    session.addConnection(frontConnection)
                }
                frontConnection.videoOrientation = .portrait
                if frontConnection.isVideoMirroringSupported {
                    frontConnection.isVideoMirrored = true
                }
            }

            // Connect back camera to back output
            if let backVideoPort = backInput.ports(for: .video, sourceDeviceType: backCamera.deviceType, sourceDevicePosition: .back).first {
                let backConnection = AVCaptureConnection(inputPorts: [backVideoPort], output: backOutput)
                if session.canAddConnection(backConnection) {
                    session.addConnection(backConnection)
                }
                backConnection.videoOrientation = .portrait
            }

            // Connect audio to both outputs
            if let audioPort = audioDeviceInput.ports(for: .audio, sourceDeviceType: audioDevice.deviceType, sourceDevicePosition: .unspecified).first {
                let frontAudioConnection = AVCaptureConnection(inputPorts: [audioPort], output: frontOutput)
                if session.canAddConnection(frontAudioConnection) {
                    session.addConnection(frontAudioConnection)
                }
            }

            session.commitConfiguration()

            // Store references
            multiCamSession = session
            frontCameraInput = frontInput
            backCameraInput = backInput
            frontVideoOutput = frontOutput
            backVideoOutput = backOutput
            audioInput = audioDeviceInput

            // Start session
            DispatchQueue.global(qos: .userInitiated).async {
                session.startRunning()
            }

            // Return preview info (texture IDs would be set up separately)
            result([
                "frontTextureId": -1, // Will use Texture widget in future
                "backTextureId": -1,
                "frontWidth": 720,
                "frontHeight": 1280,
                "backWidth": 720,
                "backHeight": 1280,
            ] as [String: Any])

        } catch {
            result(FlutterError(code: "INIT_FAILED", message: error.localizedDescription, details: nil))
        }
    }

    // MARK: - Recording

    private func startRecording(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let dir = args["sessionDir"] as? String,
              let id = args["sessionId"] as? String else {
            result(FlutterError(code: "ARGS", message: "Missing sessionDir or sessionId", details: nil))
            return
        }

        // Create session directory
        let fileManager = FileManager.default
        try? fileManager.createDirectory(atPath: dir, withIntermediateDirectories: true)

        sessionDir = dir
        sessionId = id
        recordingStartTime = Date()

        let frontPath = "\(dir)/front_camera.mp4"
        let backPath = "\(dir)/back_camera.mp4"

        // Start recording on both outputs
        frontVideoOutput?.startRecording(to: URL(fileURLWithPath: frontPath), recordingDelegate: self)
        backVideoOutput?.startRecording(to: URL(fileURLWithPath: backPath), recordingDelegate: self)

        isRecording = true

        // Start audio level monitoring
        startAudioLevelMonitoring()

        result(nil)
    }

    private func pauseRecording(result: @escaping FlutterResult) {
        frontVideoOutput?.pauseRecording()
        backVideoOutput?.pauseRecording()
        result(nil)
    }

    private func resumeRecording(result: @escaping FlutterResult) {
        frontVideoOutput?.resumeRecording()
        backVideoOutput?.resumeRecording()
        result(nil)
    }

    private func stopRecording(result: @escaping FlutterResult) {
        guard isRecording, let dir = sessionDir, let id = sessionId else {
            result(FlutterError(code: "NOT_RECORDING", message: "Not currently recording", details: nil))
            return
        }

        audioLevelTimer?.invalidate()
        audioLevelTimer = nil

        frontVideoOutput?.stopRecording()
        backVideoOutput?.stopRecording()

        isRecording = false

        let duration = Date().timeIntervalSince(recordingStartTime ?? Date())

        result([
            "sessionId": id,
            "frontVideoPath": "\(dir)/front_camera.mp4",
            "backVideoPath": "\(dir)/back_camera.mp4",
            "stereoAudioPath": "\(dir)/front_camera.mp4", // Audio embedded in front video
            "duration": duration,
        ] as [String: Any])
    }

    // MARK: - Audio Level Monitoring

    private func startAudioLevelMonitoring() {
        // Monitor audio levels every 200ms for speaker detection
        audioLevelTimer = Timer.scheduledTimer(withTimeInterval: 0.2, repeats: true) { [weak self] _ in
            guard let self = self, let sink = self.eventSink else { return }

            // Read audio levels from capture connections
            // In a full implementation, this would read from audio data output
            // For now, send simulated levels (will be replaced with real data)
            let hostLevel = self.getAudioLevel(for: self.frontVideoOutput)
            let guestLevel = self.getAudioLevel(for: self.backVideoOutput)

            sink([
                "type": "audioLevels",
                "hostLevel": hostLevel,
                "guestLevel": guestLevel,
            ] as [String: Any])
        }
    }

    private func getAudioLevel(for output: AVCaptureMovieFileOutput?) -> Double {
        guard let output = output else { return 0 }
        // Get audio channel from connections
        for connection in output.connections {
            for channel in connection.audioChannels {
                // averagePowerLevel is in dB (-160 to 0)
                let db = channel.averagePowerLevel
                // Convert to 0-1 linear scale
                let linear = max(0, (db + 60) / 60) // -60dB = 0, 0dB = 1
                return Double(linear)
            }
        }
        return 0
    }

    // MARK: - Cleanup

    private func cleanup() {
        audioLevelTimer?.invalidate()
        audioLevelTimer = nil
        multiCamSession?.stopRunning()
        multiCamSession = nil
        frontCameraInput = nil
        backCameraInput = nil
        frontVideoOutput = nil
        backVideoOutput = nil
        audioInput = nil
        isRecording = false
    }
}

// MARK: - AVCaptureFileOutputRecordingDelegate

extension DualCameraPlugin: AVCaptureFileOutputRecordingDelegate {
    func fileOutput(_ output: AVCaptureFileOutput, didFinishRecordingTo outputFileURL: URL, from connections: [AVCaptureConnection], error: Error?) {
        if let error = error {
            eventSink?([
                "type": "error",
                "message": "Recording error: \(error.localizedDescription)",
            ] as [String: Any])
        }
    }
}

// MARK: - FlutterStreamHandler

extension DualCameraPlugin: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: app/android/app/src/main/kotlin/com/stori/stori_app/DualCameraPlugin.kt
package com.stori.stori_app

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.media.MediaRecorder
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.video.*
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Dual-camera recording plugin for Android
 * Uses CameraX for concurrent front + back camera recording
 * Requires Android 11+ and device support for concurrent cameras
 */
class DualCameraPlugin private constructor(
    private val activity: Activity
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    companion object {
        private const val TAG = "DualCameraPlugin"
        private const val METHOD_CHANNEL = "com.stori/dual_camera"
        private const val EVENT_CHANNEL = "com.stori/dual_camera_events"

        fun registerWith(flutterEngine: FlutterEngine, activity: Activity) {
            val plugin = DualCameraPlugin(activity)

            val methodChannel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, METHOD_CHANNEL)
            methodChannel.setMethodCallHandler(plugin)

            val eventChannel = EventChannel(flutterEngine.dartExecutor.binaryMessenger, EVENT_CHANNEL)
            eventChannel.setStreamHandler(plugin)
        }
    }

    private var eventSink: EventChannel.EventSink? = null
    private var cameraExecutor: ExecutorService = Executors.newSingleThreadExecutor()
    private var isRecording = false
    private var sessionDir: String? = null
    private var sessionId: String? = null
    private var recordingStartTime: Long = 0
    private var isSwapped = false

    // Audio level monitoring
    private val handler = Handler(Looper.getMainLooper())
    private var audioLevelRunnable: Runnable? = null

    // CameraX components
    private var frontRecording: Recording? = null
    private var backRecording: Recording? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isDualCameraSupported" -> {
                result.success(checkDualCameraSupport())
            }
            "initialize" -> {
                initialize(call, result)
            }
            "startRecording" -> {
                startRecording(call, result)
            }
            "pauseRecording" -> {
                pauseRecording(result)
            }
            "resumeRecording" -> {
                resumeRecording(result)
            }
            "stopRecording" -> {
                stopRecording(result)
            }
            "swapCameras" -> {
                isSwapped = !isSwapped
                result.success(null)
            }
            "dispose" -> {
                cleanup()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    /**
     * Check if device supports concurrent dual cameras
     * Requires Android 11+ and hardware support
     */
    private fun checkDualCameraSupport(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) return false

        try {
            val cameraManager = activity.getSystemService(Activity.CAMERA_SERVICE) as CameraManager
            val concurrentCameraIds = cameraManager.concurrentCameraIds

            // Check if there's a set containing both front and back cameras
            for (idSet in concurrentCameraIds) {
                var hasFront = false
                var hasBack = false
                for (id in idSet) {
                    val characteristics = cameraManager.getCameraCharacteristics(id)
                    val facing = characteristics.get(CameraCharacteristics.LENS_FACING)
                    if (facing == CameraCharacteristics.LENS_FACING_FRONT) hasFront = true
                    if (facing == CameraCharacteristics.LENS_FACING_BACK) hasBack = true
                }
                if (hasFront && hasBack) return true
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error checking dual camera support", e)
        }
        return false
    }

    private fun initialize(call: MethodCall, result: MethodChannel.Result) {
        // CameraX initialization
        // In a full implementation, this sets up CameraProvider with concurrent cameras
        // For now, return placeholder values
        result.success(mapOf(
            "frontTextureId" to -1,
            "backTextureId" to -1,
            "frontWidth" to 720,
            "frontHeight" to 1280,
            "backWidth" to 720,
            "backHeight" to 1280
        ))
    }

    private fun startRecording(call: MethodCall, result: MethodChannel.Result) {
        val dir = call.argument<String>("sessionDir")
        val id = call.argument<String>("sessionId")

        if (dir == null || id == null) {
            result.error("ARGS", "Missing sessionDir or sessionId", null)
            return
        }

        // Create session directory
        File(dir).mkdirs()

        sessionDir = dir
        sessionId = id
        recordingStartTime = System.currentTimeMillis()
        isRecording = true

        // Start audio level monitoring (200ms interval)
        startAudioLevelMonitoring()

        // TODO: Start CameraX concurrent recording
        // This requires ProcessCameraProvider.getInstance() and concurrent camera use cases
        // Full implementation in Phase 3 refinement

        result.success(null)
    }

    private fun pauseRecording(result: MethodChannel.Result) {
        frontRecording?.pause()
        backRecording?.pause()
        result.success(null)
    }

    private fun resumeRecording(result: MethodChannel.Result) {
        frontRecording?.resume()
        backRecording?.resume()
        result.success(null)
    }

    private fun stopRecording(result: MethodChannel.Result) {
        if (!isRecording) {
            result.error("NOT_RECORDING", "Not currently recording", null)
            return
        }

        stopAudioLevelMonitoring()

        frontRecording?.stop()
        backRecording?.stop()

        isRecording = false

        val duration = (System.currentTimeMillis() - recordingStartTime) / 1000.0

        result.success(mapOf(
            "sessionId" to sessionId,
            "frontVideoPath" to "$sessionDir/front_camera.mp4",
            "backVideoPath" to "$sessionDir/back_camera.mp4",
            "stereoAudioPath" to "$sessionDir/front_camera.mp4",
            "duration" to duration
        ))
    }

    private fun startAudioLevelMonitoring() {
        audioLevelRunnable = object : Runnable {
            override fun run() {
                if (!isRecording) return

                // Send audio levels to Flutter
                // In full implementation, read from AudioRecord or MediaRecorder
                eventSink?.success(mapOf(
                    "type" to "audioLevels",
                    "hostLevel" to 0.0,
                    "guestLevel" to 0.0
                ))

                handler.postDelayed(this, 200)
            }
        }
        handler.post(audioLevelRunnable!!)
    }

    private fun stopAudioLevelMonitoring() {
        audioLevelRunnable?.let { handler.removeCallbacks(it) }
        audioLevelRunnable = null
    }

    private fun cleanup() {
        stopAudioLevelMonitoring()
        frontRecording?.stop()
        backRecording?.stop()
        cameraExecutor.shutdown()
        isRecording = false
    }

    // EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}
